doc/pbh/pbh-design.md

```diff
@@ -0,0 +1,758 @@
+            "priority": "1",
+            "gre_key": "0x2500/0xffffff00",
+            "inner_ether_type": "0x86dd/0xffff",
+            "hash_list": [
+                "inner_ip_proto",
+                "inner_l4_dst_port",
+                "inner_l4_src_port",
+                "inner_dst_ipv6",
+                "inner_src_ipv6"
+            ],
+            "packet_action": "SET_ECMP_HASH"
+        },
+        "pbh_table|vxlan": {
+            "priority": "2",
+            "ip_protocol": "0x11/0xff",
+            "l4_dst_port": "0x12b5/0xffff",
+            "inner_ether_type": "0x0800/0xffff",
+            "hash_list": [
+                "inner_ip_proto",
+                "inner_l4_dst_port",
+                "inner_l4_src_port",
+                "inner_dst_ipv4",
+                "inner_src_ipv4"
+            ],
+            "packet_action": "SET_LAG_HASH"
+        }
+    },
+    "PBH_TABLE": {
+        "pbh_table": {
+            "port_list": [
+                "Ethernet0",
+                "Ethernet4"
+            ],
+            "lag_list": [
+                "PortChannel0001",
+                "PortChannel0002"
+            ],
+            "description": "NVGRE and VxLAN"
+        }
+    }
+}
+```
+
+## 2.5 Flows
+
+### 2.5.1 PBH add
+
+![PBH add flow](images/pbh_add_flow.svg "Figure 3: PBH add flow")
+
+###### Figure 3: PBH add flow
+
+### 2.5.2 PBH remove
+
+![PBH remove flow](images/pbh_remove_flow.svg "Figure 4: PBH remove flow")
+
+###### Figure 4: PBH remove flow
+
+## 2.6 CLI
+
+### 2.6.1 Command structure
+
+**User interface**:
+```
+config
+|--- pbh
+     |--- table
+     |    |--- add <table_name> OPTIONS
+     |    |--- update <table_name> OPTIONS
+     |    |--- remove <table_name>
+     |
+     |--- rule
+     |    |--- add <rule_name> <table_name> OPTIONS
+     |    |--- update <rule_name> <table_name> OPTIONS
+     |    |--- remove <rule_name> <table_name>
+     |
+     |--- hash
+          |--- add <hash_name> OPTIONS
+          |--- update <hash_name> OPTIONS
+          |--- remove <hash_name>
+
+show
+|--- pbh
+     |--- table
+     |--- rule
+     |--- hash
+     |--- statistics
+```
+
+**Options:**
+
+_config pbh table add_
+1. -p|--port_list - port list
+2. -l|--lag_list - portchannel list
+3. -d|--description - table description
+
+_config pbh rule add_
+1. -p|--priority - rule priority
+2. -m|--match - match field
+3. -h|--hash_list - hash field list
+4. -a|--action=<set_ecmp_hash|set_lag_hash> - packet action
+5. -c|--counter=<enabled|disabled> - packet/byte counter
+
+_config pbh hash add_
+1. -f|--field - hash field
+3. -m|--mask - ip mask
+2. -s|--sequence - sequence id
+
+### 2.6.2 Usage examples
+
+#### 2.6.2.1 Config command group
+
+**The following command adds/updates/removes table:**
+```bash
+config pbh table add 'pbh_table' --port_list 'Ethernet0,Ethernet4' --lag_list 'PortChannel0001,PortChannel0002' \
+--description 'NVGRE and VxLAN'
+config pbh table update 'pbh_table' --port_list 'Ethernet0'
```

Comment:
I dont think, we follow "--" for any sonic CLI options today (e.g --port_list in the above command), please check, also, we can avoid table in the command.

Reply:
@venkatmahalingam we do follow `--` CLI convention in SONiC, e.g.:
```bash
root@sonic:/home/admin# config vlan member add --help
Usage: config vlan member add [OPTIONS] <vid> port

  Add VLAN member

Options:
  -u, --untagged
  -?, -h, --help  Show this message and exit.

```
As for `config pbh table add 'pbh_table'` we need an identifier (e.g., `table`) to refer to the particular item subgroup.
Not sure why you want to avoid table in the command.

Reply:
ok, that's fine, it's just a suggestion.

Reply:
@venkatmahalingam `also, we can avoid table in the command`
Please elaborate on that

Reply:
My suggestion was to have the command as follows, but I'm okay even if you keep the current command format.

config pbh add 'pbh_table' --port_list 'Ethernet0,Ethernet4' --lag_list 'PortChannel0001,PortChannel0002'

Reply:
@venkatmahalingam thanks for the clarification. Now your point is clear. Still, i am tend to keep the original CLI format. The strong argument here is: we need to keep things as simple as possible and also make sure everything is self explanatory.

Having `config pbh add 'pbh_table'` rather then `config pbh table add 'pbh_table'` is less user friendly and adds some confusion: it is not clear what object type you actually adding/updating/removing.